KNOX-2226 - Add home page to Knox

smolnar82 opened the pull request:
## What changes were proposed in this pull request?

Added a new landing - home - page for Apache Knox as well as a new REST API endpoint to query the following information (displayed on the new home page):
- general proxy information
  - Knox version
  - public certificate PEM and JKS
  - Admin UI URL
  - Admin API Details
- topologies
  - topology name
  - API/UI services (short/long description, version (if defined), Knox service URL, context path)

This new API is available via a new topology called `metadata`, whose configuration is very similar to Admin API (`admin.xml`) but there is no authorization within the topology and the only service there binds the new API to this topology.

To be able to display useful service information I enriched the existing service definitions (`service.xml` files) with the following metadata section:
```
    <metadata>
        <type></type>
        <context></context>
        <shortDesc></shortDesc>
        <description></description>
    </metadata>
```

Two new Maven modules got created:
1. `gateway-service-metadata`: contains the REST API endpoint code and related beans
2. `knox-homepage-ui`: contains the new home page front-end code (Angular2) and the required webapp related resources (`web.xml`, `gateway.xml`, `shiro.ini` where needed)

The release process has been modified too: until now the `$KNOX_DATA_DIR/deployments` folder got created the first time a topology was deployed as a web application. From now on, a static deployment comes with Knox OOTB called `homepage`. This deployment consists of two web applications: one to handle the new homepage API and another one to support the new UI.

The layout of the newly created Knox distributable looks like this:
```
├── CHANGES
├── ISSUES
├── LICENSE
├── NOTICE
├── README
├── bin
│   ├── ...
├── conf
│   ├── ...
├── data
│   ├── README
│   ├── applications
│   │   ├── README
│   │   ├── admin-ui
│   │   └── knoxauth
│   ├── deployments
│   │   ├── homepage
│   │   │   ├── %2F
│   │   │   │   └── WEB-INF
│   │   │   │       ├── gateway.xml
│   │   │   │       └── web.xml
│   │   │   └── %2Fhome
│   │   │       ├── 3rdpartylicenses.txt
│   │   │       ├── WEB-INF
│   │   │       │   ├── gateway.xml
│   │   │       │   └── web.xml
│   │   │       ├── assets
│   │   │       ├── favicon.ico
│   │   │       ├── glyphicons-halflings-regular.448c34a56d699c29117a.woff2
│   │   │       ├── glyphicons-halflings-regular.89889688147bd7575d63.svg
│   │   │       ├── glyphicons-halflings-regular.e18bbf611f2a2e43afc0.ttf
│   │   │       ├── glyphicons-halflings-regular.f4769f9bdb7466be6508.eot
│   │   │       ├── glyphicons-halflings-regular.fa2772327f55d8198301.woff
│   │   │       ├── index.html
│   │   │       ├── inline.b30c555ec93d5e8b24c8.bundle.js
│   │   │       ├── main.2089d6a43231b5fa3433.bundle.js
│   │   │       ├── main.29e6b218519192da1ad3.bundle.js
│   │   │       ├── scripts.85e9d45a06b44fd516c5.bundle.js
│   │   │       └── styles.0e0d99407f861d1ad72d.bundle.css
│   └── services
│       └── ...
├── dep
│   ├── ...
├── ext
│   ├── README
│   └── native
│       └── README
├── lib
│   ├── ...
└── samples
    └── ...
```

**The metadata API**

The new REST API's base URL is `https://$KNOX_HOST:PORT/gateway/metadata/api/v1`.
This new REST API requires basic authentication (to accommodate what we have in default topologies I choose LDAP authentication using the demo LDAP server).

- To query general proxy information the following end-point should be used:
```
https://$KNOX_HOST:PORT/gateway/metadata/api/v1/metadata/info
```

- To query _all_ topologies the following end-point should be used:
```
https://localhost:8443/gateway/metadata/api/v1/metadata/topologies
```
**Note:** by default `admin`, `manager`, `knoxsso` and `metadata` topologies are hidden, meaning they won't be added in the final result when hitting the above URL. End-users can change the list of hidden topologies by declaring `knox.homepage.hidden.topologies` configuration in `gateway-site.xml`. This is configuration must hold a comma-separated list of topology names if defined.

- To query a particular - non-hidden - topology the following end-point should be used:
```
https://localhost:8443/gateway/metadata/api/v1/metadata/topologies/$topologyName
```

**The homepage UI**

The Knox Home page is available here: https://$KNOX_HOST:PORT/home.

Consists of two sections: general proxy information and topology browsing (see a detailed list of displayed items above). The UI uses `SSOCookieFederationFilter` to authenticate trough KnoxSSO. 

## How was this patch tested?

JUnit:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21:19 min (Wall Clock)
[INFO] Finished at: 2020-02-13T23:05:12+01:00
[INFO] Final Memory: 431M/2205M
[INFO] ------------------------------------------------------------------------
```

I executed a comprehensive manual test to see if both the new API and the home page works as expected.

Fetching general proxy information:
```
$ curl -iku admin:admin-password -H "Accept: application/json" 'https://localhost:8443/gateway/metadata/api/v1/metadata/info'
HTTP/1.1 200 OK
Date: Mon, 24 Feb 2020 19:56:19 GMT
Set-Cookie: KNOXSESSIONID=node0sx82cct3pbunsvxaodv3g8872.node0; Path=/gateway/metadata; Secure; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/gateway/metadata; Max-Age=0; Expires=Sun, 23-Feb-2020 19:56:19 GMT
Content-Type: application/json
Content-Length: 412
Server: Jetty(9.4.26.v20200117)

{
   "generalProxyInfo" : {
      "version" : "1.4.0-SNAPSHOT (hash=ddb0868834a7fb283ab11d0fb161d68bcd884eee)",
      "publicCertPemPath" : "assets/gateway-client-trust.pem",
      "publicCertJksPath" : "assets/gateway-client-trust.jks",
      "adminUiUrl" : "https://localhost:8443/gateway/manager/admin-ui/",
      "adminApiBookUrl" : "https://knox.apache.org/books/knox-1-4-0/user-guide.html#Admin+API"
   }
}
```

Fetching all (non-hidden) topologies:
```
$ curl -iku admin:admin-password -H "Accept: application/json" 'https://localhost:8443/gateway/metadata/api/v1/metadata/topologies'
HTTP/1.1 200 OK
Date: Mon, 24 Feb 2020 19:57:17 GMT
Set-Cookie: KNOXSESSIONID=node0asbc4oe56dg71o036vanuwfl3.node0; Path=/gateway/metadata; Secure; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/gateway/metadata; Max-Age=0; Expires=Sun, 23-Feb-2020 19:57:17 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Server: Jetty(9.4.26.v20200117)

{
   "topologyInformations" : {
      "topologyInformation" : [ {
         "topology" : "cdp-proxy",
         "apiServices" : {
            "service" : [ {
               "context" : "/atlas/api/",
               "description" : "Atlas exposes a variety of REST endpoints to work with types, entities, lineage and data discovery.",
               "serviceName" : "ATLAS-API",
               "serviceUrl" : "https://localhost:8443/gateway/cdp-proxy/atlas/api/",
               "shortDesc" : "Atlas API",
               "type" : "API"
            }, {
            ...
               "context" : "/zeppelin/ws/",
               "description" : "Apache Zeppelin is a web-based notebook that enables data-driven, interactive data analytics and collaborative documents with SQL, Scala and more.",
               "serviceName" : "ZEPPELINWS",
               "serviceUrl" : "https://localhost:8443/gateway/cdp-proxy/zeppelin/ws/",
               "shortDesc" : "Zeppelin WS API",
               "type" : "API"
            } ]
         },
         "uiServices" : {
            "service" : [ {
               "context" : "/atlas/",
               "description" : "Apache Atlas provides open metadata management and governance capabilities for organizations to build a catalog of their data assets, \n            classify and govern these assets and provide collaboration capabilities around these data assets for data scientists, analysts and the data governance team.",
               "serviceName" : "ATLAS",
               "serviceUrl" : "https://localhost:8443/gateway/cdp-proxy/atlas/",
               "shortDesc" : "Atlas UI",
               "type" : "UI"
            }, {
               "context" : "/druid-coordinator-ui/",
               "description" : "The Druid Coordinator exposes a web GUI for displaying cluster information and rule configuration.",
               "serviceName" : "DRUID-COORDINATOR-UI",
               "serviceUrl" : "https://localhost:8443/gateway/cdp-proxy/druid-coordinator-ui/",
               "shortDesc" : "Druid Coordinator Console",
               "type" : "UI"
            }, {
            ...
            
               "context" : "/yarnuiv2/",
               "description" : "YARN Resource Manager Web interface (v2)",
               "serviceName" : "YARNUIV2",
               "serviceUrl" : "https://localhost:8443/gateway/cdp-proxy/yarnuiv2/",
               "shortDesc" : "YARN Resource Manager Web UI V2",
               "type" : "UI"
            }, {
               "context" : "/zeppelin/",
               "description" : "Zeppelin's web user interface allows end-users to explore Zeppelin notebooks.",
               "serviceName" : "ZEPPELINUI",
               "serviceUrl" : "https://localhost:8443/gateway/cdp-proxy/zeppelin/",
               "shortDesc" : "Zeppelin Web UI",
               "type" : "UI"
            } ]
         }
      }, {
         "topology" : "sandbox",
         "apiServices" : {
            "service" : [ {
            ...
            } ]
         },
         "uiServices" : {
            "service" : [ {
               "context" : "/druid-coordinator-ui/",
               "description" : "The Druid Coordinator exposes a web GUI for displaying cluster information and rule configuration.",
               "serviceName" : "DRUID-COORDINATOR-UI",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-coordinator-ui/",
               "shortDesc" : "Druid Coordinator Console",
               "type" : "UI"
            }, {
               "context" : "/druid-overlord-ui/",
               "description" : "The Overlord provides a UI for managing tasks and workers.",
               "serviceName" : "DRUID-OVERLORD-UI",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-overlord-ui/",
               "shortDesc" : "Druid Overlord Console",
               "type" : "UI"
            }, {
               "context" : "/hue/",
               "description" : "Hue UI is a Web interface for analyzing data with Apache Hadoop",
               "serviceName" : "HUE",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/hue/",
               "shortDesc" : "Hue UI",
               "type" : "UI"
            } ]
         }
      } ]
   }
}
```

Fetching `sandbox` topology:
```
$ curl -iku admin:admin-password -H "Accept: application/json" 'https://localhost:8443/gateway/metadata/api/v1/metadata/topologies/sandbox'
HTTP/1.1 200 OK
Date: Mon, 24 Feb 2020 19:58:26 GMT
Set-Cookie: KNOXSESSIONID=node01vbyvl5o66n1t7a56eh32wt4v4.node0; Path=/gateway/metadata; Secure; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/gateway/metadata; Max-Age=0; Expires=Sun, 23-Feb-2020 19:58:26 GMT
Content-Type: application/json
Content-Length: 6645
Server: Jetty(9.4.26.v20200117)

{
   "topologyInformations" : {
      "topologyInformation" : [ {
         "topology" : "sandbox",
         "apiServices" : {
            "service" : [ {
               "context" : "/namenode/",
               "description" : "Namenode",
               "serviceName" : "NAMENODE",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/namenode/",
               "shortDesc" : "Namenode",
               "type" : "UNKNOWN"
            }, {
               "context" : "/jobtracker/",
               "description" : "Jobtracker",
               "serviceName" : "JOBTRACKER",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/jobtracker/",
               "shortDesc" : "Jobtracker",
               "type" : "UNKNOWN"
            }, {
               "context" : "/webhdfs/",
               "description" : "An HTTP REST API which supports the complete FileSystem interface for HDFS.",
               "serviceName" : "WEBHDFS",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/webhdfs/",
               "shortDesc" : "Web HDFS",
               "type" : "API"
            }, {
               "context" : "/templeton/",
               "description" : "WebHCat is the REST API for HCatalog, a table and storage management layer for Hadoop.",
               "serviceName" : "WEBHCAT",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/templeton/",
               "shortDesc" : "WebHCat",
               "type" : "API"
            }, {
               "context" : "/oozie/",
               "description" : "Oozie is a workflow scheduler system to manage Apache Hadoop jobs. The Oozie Web Services API is a HTTP REST JSON API.",
               "serviceName" : "OOZIE",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/oozie/",
               "shortDesc" : "Oozie Web Services API",
               "type" : "API"
            }, {
               "context" : "/hbase/",
               "description" : "The HBase REST server exposes endpoints that provide CRUD (create, read, update, delete) operations for each HBase process, as well as tables, regions, and namespaces.",
               "serviceName" : "WEBHBASE",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/hbase/",
               "shortDesc" : "Web HBase",
               "type" : "API"
            }, {
               "context" : "/hive/",
               "description" : "The Apache Hive ™ data warehouse software facilitates reading, writing, and managing large datasets residing in distributed storage using SQL. Structure can be projected onto data already in storage. A command line tool and JDBC driver are provided to connect users to Hive.",
               "serviceName" : "HIVE",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/hive/",
               "shortDesc" : "Hive",
               "type" : "API"
            }, {
               "context" : "/resourcemanager/",
               "description" : "The YARN Resource Manager Service (RM) is the central controlling authority for resource management and makes allocation decisions ResourceManager has two main components: Scheduler and ApplicationsManager.",
               "serviceName" : "RESOURCEMANAGER",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/resourcemanager/",
               "shortDesc" : "YARN Resource Manager",
               "type" : "API"
            }, {
               "context" : "/druid-coordinator/",
               "description" : "The Druid Coordinator process is primarily responsible for segment management and distribution. \n            More specifically, the Druid Coordinator process communicates to Historical processes to load or drop segments based on configurations. \n            The Druid Coordinator is responsible for loading new segments, dropping outdated segments, managing segment replication, and balancing segment load.",
               "serviceName" : "DRUID-COORDINATOR",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-coordinator/",
               "shortDesc" : "Druid Coordinator API",
               "type" : "API"
            }, {
               "context" : "/druid-broker/druid/",
               "description" : "The Broker is the process to route queries to if you want to run a distributed cluster. \n            It understands the metadata published to ZooKeeper about what segments exist on what processes and routes queries such that they hit the right processes. \n            This process also merges the result sets from all of the individual processes together. On start up, Historical processes announce themselves and the segments they are serving in Zookeeper.",
               "serviceName" : "DRUID-BROKER",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-broker/druid/",
               "shortDesc" : "Druid Broker API",
               "type" : "API"
            }, {
               "context" : "/druid-router/",
               "description" : "The Apache Druid Router process can be used to route queries to different Broker processes. By default, the broker routes queries based on how Rules are set up",
               "serviceName" : "DRUID-ROUTER",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-router/",
               "shortDesc" : "Druid Router API",
               "type" : "API"
            }, {
               "context" : "/druid-overlord/",
               "description" : "The Overlord process is responsible for accepting tasks, coordinating task distribution, creating locks around tasks, and returning statuses to callers. \n            Overlord can be configured to run in one of two modes - local or remote (local being default).",
               "serviceName" : "DRUID-OVERLORD",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-overlord/",
               "shortDesc" : "Druid Overlord API",
               "type" : "API"
            } ]
         },
         "uiServices" : {
            "service" : [ {
               "context" : "/druid-coordinator-ui/",
               "description" : "The Druid Coordinator exposes a web GUI for displaying cluster information and rule configuration.",
               "serviceName" : "DRUID-COORDINATOR-UI",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-coordinator-ui/",
               "shortDesc" : "Druid Coordinator Console",
               "type" : "UI"
            }, {
               "context" : "/druid-overlord-ui/",
               "description" : "The Overlord provides a UI for managing tasks and workers.",
               "serviceName" : "DRUID-OVERLORD-UI",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/druid-overlord-ui/",
               "shortDesc" : "Druid Overlord Console",
               "type" : "UI"
            }, {
               "context" : "/hue/",
               "description" : "Hue UI is a Web interface for analyzing data with Apache Hadoop",
               "serviceName" : "HUE",
               "serviceUrl" : "https://localhost:8443/gateway/sandbox/hue/",
               "shortDesc" : "Hue UI",
               "type" : "UI"
            } ]
         }
      } ]
   }
}
```

Logged in to the homepage via the KnoxSSO login page:
<img width="1052" alt="Screen Shot 2020-02-13 at 11 08 08 PM" src="https://user-images.githubusercontent.com/34065904/74482976-d05ca100-4eb5-11ea-9245-38241a406247.png">


Expanded `cdp-proxy` in `Topology Browsing`:
<img width="1680" alt="Screen Shot 2020-02-26 at 10 55 56 AM" src="https://user-images.githubusercontent.com/34065904/75333548-99f42e00-5886-11ea-8942-a60001b739f6.png">

risdenk commented on gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java:
```diff
@@ -923,6 +925,40 @@ private synchronized void internalActivateArchive( Topology topology, File warDi
+      // API
+      activateHomepageContext("/home/api", "homepage/%2F");
+    } catch (Throwable e) {
+      e.printStackTrace();
```
remove `e.printStackTrace()`

why are we catching `Throwable` that seems bad

smolnar82 replied:
Deleted.

risdenk commented on gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java:
```diff
@@ -923,6 +925,40 @@ private synchronized void internalActivateArchive( Topology topology, File warDi
+  private synchronized void activateHomeApp() {
+    try {
+      // UI
+      activateHomepageContext("/home", "homepage/%2Fhome");
```
why the weird `%2F` - which is a url encoded forward slash...

smolnar82 replied:
I've just followed the process which is already in place for dynamic topology deployment. That process also creates `%2F` and `%2Fadmin-ui` folders where necessary. When tested, it was not even working w/o these prefixes.

risdenk commented on knox-homepage-backend/src/main/java/org/apache/knox/homepage/GeneralProxyInformationMarshaller.java:
```diff
@@ -0,0 +1,89 @@
+    return MediaType.APPLICATION_JSON_TYPE.getSubtype().equals(mediaType.getSubtype()) ? getJsonMarshaller() : getXmlMarshaller();
+  }
+
+  private synchronized Marshaller getXmlMarshaller() throws JAXBException {
```
Need to make sure this is a safe marshaller and we aren't open to XML issues. I thought we had a class to handle a lot of this complexity.

smolnar82 replied:
This is the same pattern we have in `gateway-admin-ui`. I agree this should be enhanced but I believe it's out of the scope of this JIRA.

risdenk commented on knox-homepage-backend/src/main/java/org/apache/knox/homepage/HomePageResource.java:
```diff
@@ -0,0 +1,126 @@
+        proxyInfo.setPublicCertJksPath("assets/gateway-client-trust.jks");
+        proxyInfo.setAdminUiUrl(getBaseGatewayUrl(config) + "/manager/admin-ui/");
+      }
+    } catch (Exception e) {
```
Ummm should do something here...

smolnar82 replied:
Fixed.

risdenk commented on knox-homepage-backend/src/main/java/org/apache/knox/homepage/TopologyInformationWrapperMarshaller.java:
```diff
@@ -0,0 +1,89 @@
+    return MediaType.APPLICATION_JSON_TYPE.getSubtype().equals(mediaType.getSubtype()) ? getJsonMarshaller() : getXmlMarshaller();
+  }
+
+  private synchronized Marshaller getXmlMarshaller() throws JAXBException {
```
Need to make sure this is a safe marshaller and we aren't open to XML issues. I thought we had a class to handle a lot of this complexity.

smolnar82 replied:
This is the same pattern we have in `gateway-admin-ui`. I agree this should be enhanced but I believe it's out of the scope of this JIRA.

risdenk commented on gateway-release/src/assembly.xml:
```diff
@@ -90,6 +90,13 @@
+            <directory>${project.build.directory}/deployments</directory>
+            <outputDirectory>data/deployments</outputDirectory>
+            <includes>
+                <include>**/%2F*/**/*</include>
```
weird `%2F`??

smolnar82 replied:
There are two additional folders here (`META-INF` and the war-plugin generated empty `WEB-INF` which should not be placed in the final `deployments` folder.

risdenk commented on gateway-release/pom.xml:
```diff
@@ -144,6 +144,14 @@
                                             <outputDirectory>${project.build.directory}</outputDirectory>
                                             <includes>applications/**/*</includes>
                                         </artifactItem>
+                                        <artifactItem>
```
Need to see what this looks like in the final tar.gz to make sure this is valid packaging since `gateway-release` basically builds the final apache builds.

smolnar82 replied:
I'll add this info to the PR description later today.

smolnar82 replied:
Done

risdenk commented on knox-homepage-ui/pom.xml:
```diff
@@ -0,0 +1,115 @@
+    <description>The Knox Home Page UI</description>
+
+    <properties>
+        <frontend-maven-plugin.version>1.9.1</frontend-maven-plugin.version>
```
Would be good to move these 2 properties to top level pom since now its duplicated w/ `gateway-admin-ui/pom.xml`

smolnar82 replied:
Done.

risdenk commented on knox-homepage-ui/homepage/assets/vkbeautify.js:
```diff
@@ -0,0 +1,392 @@
+/*
```
I'd personally love to get rid of this file both here and in gateway-admin-ui. Either its something we can pull in via npm or we should have it minified. Right now its weird we straight copy this around.

smolnar82 replied:
Done in both projects (pulled in by installing with npm)

risdenk commented on knox-homepage-ui/homepage/app/generalProxyInformation/general.proxy.information.component.ts:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    getAdminApiUrl() {
+        return 'https://knox.apache.org/books/knox-1-4-0/user-guide.html#Admin+API';
```
should just be `https://knox.apache.org/books/knox-1-4-0/user-guide.html#Admin+API` Looks like smart dashes in the line.

In either case since we are printing the version of Knox - it should be possible to dynamically created this url based on `this.generalProxyInformation.version`

smolnar82 replied:
Done: admin API book URL is calculated on the back-end - based on the build version - and being sent as part of the general proxy information payload.

risdenk commented on knox-homepage-ui/angular-cli.json:
```diff
@@ -0,0 +1,56 @@
+{
+  "project": {
+    "version": "1.0.0-beta.22-1",
```
where is this version from? :D

smolnar82 replied:
Changed

risdenk commented on knox-homepage-ui/angular-cli.json:
```diff
@@ -0,0 +1,56 @@
+      "scripts": [
+        "../node_modules/jquery/dist/jquery.min.js",
+        "../node_modules/bootstrap/dist/js/bootstrap.js",
+        "../node_modules/ace-builds/src-min/ace.js",
```
Do we need `ace` since I think thats the editor.

smolnar82 replied:
Removed all that not needed.

smolnar82 commented on gateway-service-definitions/src/main/resources/services/cm-ui/rewrite.xml:
```diff
@@ -0,0 +1,33 @@
+<!--
```
Nice catch; fixed!

smolnar82 commented on gateway-service-definitions/src/main/resources/services/cm-ui/service.xml:
```diff
@@ -0,0 +1,62 @@
+<!--
```
Nice catch; fixed!

smolnar82 commented on gateway-service-definitions/src/main/resources/services/hdfsui/3.0.0/service.xml:
```diff
@@ -16,6 +16,12 @@
 <service role="HDFSUI" name="hdfs" version="3.0.0">
+    <metadata>
+        <type>UI</type>
+        <context>/hdfs?host=%BACKEND_HOST</context>
```
Done.

risdenk commented on gateway-service-definitions/src/main/resources/services/hdfsui/2.7.0/service.xml:
```diff
@@ -16,6 +16,12 @@
 <service role="HDFSUI" name="hdfs" version="2.7.0">
+    <metadata>
+        <type>UI</type>
+        <context>/hdfs?host={{BACKEND_HOST}}</context>
```
There is no `?host=` parameter on the 2.7.0 version of the HDFS UI service definition.

smolnar82 replied:
Fixed.

risdenk commented on gateway-service-definitions/src/main/resources/services/hdfsui/3.0.0/service.xml:
```diff
@@ -16,6 +16,12 @@
 <service role="HDFSUI" name="hdfs" version="3.0.0">
+    <metadata>
+        <type>UI</type>
+        <context>/hdfs?host={{BACKEND_HOST}}</context>
```
should be `/hdfs/?host=` notice the trailing slash after `/hdfs`

smolnar82 replied:
Fixed.

risdenk commented on knox-homepage-backend/pom.xml:
```diff
@@ -0,0 +1,67 @@
+
+    <artifactId>knox-homepage-backend</artifactId>
+    <name>knox-homepage-backend</name>
+    <description>The Knox Home Page BE</description>
```
`BE`?

smolnar82 replied:
Fixed :)

risdenk commented on knox-homepage-backend/src/main/java/org/apache/knox/homepage/HomePageResource.java:
```diff
@@ -0,0 +1,197 @@
+
+  @GET
+  @Produces({ APPLICATION_JSON, APPLICATION_XML })
+  @Path("generalProxyInformation")
```
How about just `info` instead of `generalProxyInformation`? Something simpler.

smolnar82 replied:
Done.

risdenk commented:
One other thing to maybe think about is how easy it is to change the homepage authentication/authorization. Right now we end up with separate Shiro and configs. This is a bit clunky especially when demo ldap won't exist for most users when they deploy for real.

smolnar82 commented:
@risdenk ,

WRT your JSON question: on one hand, you are right, it'd be cleaner when reading the JSON as you recommended. On the other hand, this API generates XML output too, and in that case, the current structure is cleaner IMO:
```
      <uiServices>
         <service>
            <context>/druid-coordinator-ui/</context>
            <description>The Druid Coordinator exposes a web GUI for displaying cluster information and rule configuration.</description>
            <serviceName>DRUID-COORDINATOR-UI</serviceName>
            <serviceUrl>https://localhost:8443/gateway/sandbox/druid-coordinator-ui/</serviceUrl>
            <shortDesc>Druid Coordinator Console</shortDesc>
            <type>UI</type>
         </service>
         <service>
            <context>/druid-overlord-ui/</context>
            <description>The Overlord provides a UI for managing tasks and workers.</description>
            <serviceName>DRUID-OVERLORD-UI</serviceName>
            <serviceUrl>https://localhost:8443/gateway/sandbox/druid-overlord-ui/</serviceUrl>
            <shortDesc>Druid Overlord Console</shortDesc>
            <type>UI</type>
         </service>
         <service>
            <context>/hue/</context>
            <description>Hue UI is a Web interface for analyzing data with Apache Hadoop</description>
            <serviceName>HUE</serviceName>
            <serviceUrl>https://localhost:8443/gateway/sandbox/hue/</serviceUrl>
            <shortDesc>Hue UI</shortDesc>
            <type>UI</type>
         </service>
      </uiServices>
```

Moreover, the current solution is consistent with Admin API outputs:
```
$ curl -iku admin:admin-password -H "Accept: application/json" 'https://localhost:8443/gateway/admin/api/v1/topologies/'
HTTP/1.1 200 OK
Date: Fri, 14 Feb 2020 08:26:44 GMT
Set-Cookie: KNOXSESSIONID=node01wffn263vit5nemdc9021f45z3.node0; Path=/gateway/admin; Secure; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/gateway/admin; Max-Age=0; Expires=Thu, 13-Feb-2020 08:26:44 GMT
Content-Type: application/json
Content-Length: 1160
Server: Jetty(9.4.26.v20200117)

{
   "topologies" : {
      "topology" : [ {
         "name" : "admin",
         "timestamp" : "1581667994000",
         "uri" : "https://localhost:8443/gateway/admin",
         "href" : "https://localhost:8443/gateway/admin/api/v1/topologies/admin"
      }, {
         "name" : "cdp-proxy",
         "timestamp" : "1581625805000",
         "uri" : "https://localhost:8443/gateway/cdp-proxy",
         "href" : "https://localhost:8443/gateway/admin/api/v1/topologies/cdp-proxy"
      }, {
         "name" : "knoxsso",
         "timestamp" : "1581625577000",
         "uri" : "https://localhost:8443/gateway/knoxsso",
         "href" : "https://localhost:8443/gateway/admin/api/v1/topologies/knoxsso"
      }, {
         "name" : "manager",
         "timestamp" : "1581667994000",
         "uri" : "https://localhost:8443/gateway/manager",
         "href" : "https://localhost:8443/gateway/admin/api/v1/topologies/manager"
      }, {
         "name" : "sandbox",
         "timestamp" : "1581625577000",
         "uri" : "https://localhost:8443/gateway/sandbox",
         "href" : "https://localhost:8443/gateway/admin/api/v1/topologies/sandbox"
      } ]
   }
}
```

risdenk commented:
Thanks for the explanation of the XML vs JSON thing. Super confusing but see how it ends up that way.

smolnar82 commented:
So I added another commit to clear out the authentication stuff:
- the homepage UI uses `knoxsso` and deploys the new API under `/homepage/` the same way `manager` deploys the admin API under `/manager`. This way there is no need to login twice and `knoxsso` is configurable easily.
- the new API is deployed through a new topology called `metadata`. This topology is configured to use LDAP authentication (to be consistent with Admin API). I'm going to modify the PR description accordingly.

risdenk commented on gateway-provider-security-jwt/src/main/java/org/apache/knox/gateway/provider/federation/jwt/filter/SSOCookieFederationFilter.java:
```diff
@@ -93,6 +95,13 @@ public void init( FilterConfig filterConfig ) throws ServletException {
+      if (gatewayConfig != null) {
+        gatewayPath = gatewayConfig.getGatewayPath();
+      }
+    }
```
This should be its own method

risdenk replied:
Do we need `StringUtils.isBlank`? Check is not null and is not empty? That would remove the commons-lang3 dependency

smolnar82 replied:
I know you are not fan of adding new dependencies but the `commons-lang` dependency is in our project anyway. If you are insisting on removing it I'll do...

smolnar82 replied:
> This should be its own method

+1 (let me wait a bit until for review comments before submitting a new patchset)

smolnar82 replied:
Fixed:
- setting `gatewayPath` in its own method
- removed the `commons-lang3` dependency

risdenk commented on gateway-release/home/conf/topologies/metadata.xml:
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Is this topology just an example? Does it need to be exposed by default? If its an example `gateway-release/home/templates/` would be a better place.

smolnar82 replied:
No, it's not. Without this topology, clients (CM, Ambari, etc...) can not query metadata (general proxy info and topology information).
Please note this can not be added into Admin API because there is no need to authorize the incoming requests.

lmccay replied:
I see that the metadata.xml topology is aligned with other existing topologies like admin.xml for instance. I think that we need to consider moving these to descriptors that point to the default-providers.json shared provider config. This way the authentication can be changed in one place. Admin will require an authorization provider but since it is for the admin service and not others that should be okay. Let's not block this PR and follow up with another to move to descriptors. Just wanted to mention it here.

smolnar82 replied:
https://issues.apache.org/jira/browse/KNOX-2251

smolnar82 replied:
https://issues.apache.org/jira/browse/KNOX-2251

pzampino commented on knox-homepage-backend/pom.xml:
```diff
@@ -0,0 +1,59 @@
+        <version>1.4.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>knox-homepage-backend</artifactId>
```
Why are these new modules prefixed with "knox" rather than "gateway"?

smolnar82 replied:
I think it expresses more that they belong to Knox as a product, including the gateway, the CLI, the KnoxShell rather than referring only the gateway part.

pzampino replied:
While some of the information supports the client pieces, I still feel that the information is all about the gateway. I could be persuaded otherwise, but this is my feeling at this point.

smolnar82 replied:
The idea is to add `KnoxShell/Client Downloads and docs` later on into the `General Proxy Information` block, but it's been scoped out for now.
I gave a second thought about the module names and I think the UI should be kept as-is (`knox-homepage-ui`) because of the `General Proxy Information` block. However, the back-end module can - and should - be renamed like `gateway-service-metadata`.
What do you think?

pzampino replied:
Sounds good to me. I was unaware of the client downloads bit.

smolnar82 replied:
Renamed the module as discussed above.

moresandeep commented on gateway-service-definitions/src/main/resources/services/cm-ui/1.0.0/rewrite.xml:
```diff
@@ -0,0 +1,33 @@
+  <rule dir="IN" name="CM-UI/cm-ui/ui/root" pattern="*://*:*/**/cmf/">
+    <rewrite template="{$serviceUrl[CM-UI]}/cmf/"/>
+  </rule>
+  <rule dir="IN" name="CM-UI/cm-ui/static/inbound" pattern="*://*:*/**/static/{xxx_unique_path=**}?{**}">
```
why is the context path /static here shouldn't it be /cmf/static ?

smolnar82 replied:
As discussed this is OK.

moresandeep commented on gateway-service-definitions/src/main/resources/services/cm-ui/1.0.0/rewrite.xml:
```diff
@@ -0,0 +1,33 @@
+  <rule dir="IN" name="CM-UI/cm-ui/static/inbound" pattern="*://*:*/**/static/{xxx_unique_path=**}?{**}">
+    <rewrite template="{$serviceUrl[CM-UI]}/static/{xxx_unique_path=**}?{**}"/>
+  </rule>
+  <rule dir="IN" name="CM-UI/cm-ui/rest" pattern="*://*:*/**/api/{xxx_unique_path=**}?{**}">
```
Same as above.

smolnar82 replied:
As discussed this is OK.

moresandeep commented on gateway-service-definitions/src/main/resources/services/cm-ui/1.0.0/rewrite.xml:
```diff
@@ -0,0 +1,33 @@
+  <rule dir="IN" name="CM-UI/cm-ui/rest" pattern="*://*:*/**/api/{xxx_unique_path=**}?{**}">
+    <rewrite template="{$serviceUrl[CM-UI]}/api/{xxx_unique_path=**}?{**}"/>
+  </rule>
+  <rule dir="IN" name="CM-UI/cm-ui/root" pattern="*://*:*/*/*/{xxx_unique_path=*}?{**}">
```
same as above

smolnar82 replied:
As discussed this is OK.

smolnar82 commented on gateway-service-definitions/src/main/resources/services/ambari/0.2.2.0/service.xml:
```diff
@@ -15,6 +15,12 @@
 <service role="AMBARI" name="ambari" version="0.2.2.0">
+    <metadata>
+        <type>API</type>
+        <context>/ambari/api</context>
```
I know it's redundant but it's easier to add it in the metadata than parsing the `route` elements.
It's useful to know what comes after `$KNOX_URL/$GATEWAY_PATH/`

moresandeep replied:
Created a JIRA so we can use `context` in routes 
https://issues.apache.org/jira/browse/KNOX-2252

smolnar82 commented on knox-homepage-backend/src/main/java/org/apache/knox/homepage/service/model/ServiceModel.java:
```diff
@@ -0,0 +1,89 @@
+  public void setService(Service service) {
+    this.service = service;
+  }
+
```
+1

smolnar82 replied:
Done.

risdenk commented:
So some feedback after spinning this up and looking at it:
* why is it at /home instead of just at /? ie: https://localhost:8443 instead of requiring a path?
* There is some weird spacing going on - there should be a reasonable border around the page?

![Screen Shot 2020-02-26 at 5 02 22 PM](https://user-images.githubusercontent.com/3384157/75392359-de4cf180-58b9-11ea-95ea-bd8407701b9d.png)

* Expanding the services there is an "i" button that does nothing?

* Hitting https://localhost:8443/gateway/metadata/api/v1/metadata/info doesn't return a way to download the pem/jks certs. 
** Can't the full paths be exposed?
** Can't the cert be on its own API endpoint to download?
```
<publicCertPemPath>assets/gateway-client-trust.pem</publicCertPemPath>
<publicCertJksPath>assets/gateway-client-trust.jks</publicCertJksPath>
```

* Hive has a `<serviceUrl>https://localhost:8443/gateway/sandbox/hive/</serviceUrl>` url instead of a proper JDBC url?

smolnar82 commented:
My latest changes include:
- appropriate bordering on homepage
- homepage is bound to `/` (i.e. is available on https://localhost:8443)
- new metadata API endpoint to download public certificate in PEM/JKS format
- changed Hive's service URL to a sample JDBC URL
- re-ordered API/UI services on homepage
- renamed `Topology Browsing` to `Topologies`
- `General Proxy Information` font size is the same as `Topologies`
- new `Metadata API` links are added into `General Proxy Information`

@risdenk, the (i) buttons are tooltips; they show a meaningful description of the service

I'm going to file a follow-up JIRA to enhance the homepage based on @lmccay 's recommendations

risdenk commented on knox-homepage-ui/homepage/app/generalProxyInformation/general.proxy.information.component.ts:
```diff
@@ -0,0 +1,113 @@
+                                <a href="{{ getMetadataAPIUrl('info') }}" target="_blank">General Proxy Information</a>
+                                &nbsp;&nbsp;|&nbsp;&nbsp;
+                                <a href="{{ getMetadataAPIUrl('topologies') }}" target="_blank">Topologies</a>
+                            </td>
```
nit: I think in theory this can be in a separate file?

smolnar82 replied:
Done

risdenk commented on gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/ServiceModel.java:
```diff
@@ -0,0 +1,121 @@
+          request.getServerPort(), gatewayPath, topologyName, context);
+    } else {
+      if (context.indexOf("{{BACKEND_HOST}}") > -1) {
+        context = context.replaceAll("\\{\\{BACKEND_HOST\\}\\}", getBackendServiceUrl());
```
I don't think you need `replaceAll` here just `replace` since `replaceAll` is for regex.

smolnar82 replied:
Fixed.

risdenk commented on gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/ServiceModel.java:
```diff
@@ -0,0 +1,121 @@
+  @XmlElement
+  public String getServiceUrl() {
+    String context = getContext();
+    if ("HIVE".equals(getServiceName())) {
```
Long term I hope we don't have to do this in the ServiceModel? It would be better to have this captured in the service metadata.

smolnar82 commented:
Follow-up JIRA: https://issues.apache.org/jira/browse/KNOX-2254
